### PR TITLE
combine the yurthub's server and client certificate into one

### PR DESCRIPTION
**Pull request by qclc**
/kind enhancement

#### What this PR does / why we need it:
1. **Combine the yurthub server certificate and client certificate into one to reduce the management and maintenance cost of multiple certificates:**
    - The new certificate is stored in `/var/lib/yurthub/pki/yurthub-current.pem` by default and will replace the old yurthub client certificate.
2. **The new yurthub certificate mode is compatible with the old version**:
    - After yurthub is deployed again, if yurthub detects that the old certificate is inconsistent with the certificate currently requested (for example, the server key usage is not included), certificate rotation will be automatically triggered to obtain a new certificate.
    - Yurthub can still use the old server certificate (yurthub-server-current.pem) to provide server services during the rotation period.
 3. **Add a new yurthub flag named `--hub-cert-organizations`** :
    - It allows user to add additional Organizations to their yurthub certificates.
#### Which issue(s) this PR fixes:

Fixes #

#### Special notes for your reviewer:

/assign @rambohe-ch please take review

#### Does this PR introduce a user-facing change?

Add a new yurthub flag named `--hub-cert-organizations`. It allows user to add additional Organizations to their yurthub certificates.

### CODE REVIEW

File: pkg/yurthub/certificate/pki.go [hunk 1]
```diff
@@ -83,3 +101,43 @@ func GenCertPoolUseCA(caFile string) (*x509.CertPool, error) {
 	certPool.AppendCertsFromPEM(caData)
 	return certPool, nil
 }
+
+// CheckValidityOfCertificate checks whether the certificate is valid or not
+func CheckValidityOfCertificate(cert *tls.Certificate) bool {
```
Comment (rambohe-ch) [inline pkg/yurthub/certificate/pki.go:106]: how about modify `CheckValidityOfCertificate` to `hasServerAuthUsage`?

**Comment by rambohe-ch** [comment]
@qclc please fix unit test error.

**Comment by qclc** [comment]
> @qclc please fix unit test error.

The error has fixed, please take review.

File: pkg/yurthub/server/server.go [hunk 2]
```diff
@@ -169,45 +163,18 @@ func healthz(w http.ResponseWriter, _ *http.Request) {
...
-	}
-	// create a certificate manager for the yurthub server and run the csr approver for both yurthub
-	serverCertMgr, err := server.NewYurtHubServerCertManager(clientSet, certDir, proxyServerSecureDummyAddr)
-	if err != nil {
-		return nil, err
-	}
-	serverCertMgr.Start()
-
+// GenUseCertMgrAndTLSConfig generate a TLS configuration by using the hub certificate manager(which is used for server and client).
+func GenUseCertMgrAndTLSConfig(certificateMgr interfaces.YurtCertificateManager, certDir string, stopCh <-chan struct{}) (*tls.Config, error) {
```
[the remark this answers is not in the document]
Comment (qclc) [reply pkg/yurthub/server/server.go]: The unused stopCh has been deleted, please take review.

File: pkg/yurthub/certificate/pki.go (hunk 1 above)
Comment (qclc) [reply pkg/yurthub/certificate/pki.go]: The function name has been modified, please take review.

File: pkg/yurthub/certificate/pki.go [hunk 3]
```diff
@@ -41,22 +103,9 @@ func GenTLSConfigUseCertMgrAndCertPool(
...
-	tlsConfig.GetCertificate =
-		func(*tls.ClientHelloInfo) (*tls.Certificate, error) {
-			cert := m.Current()
-			if cert == nil {
-				return &tls.Certificate{Certificate: nil}, nil
-			}
-			return cert, nil
-		}
+	scm := newHubServerCertMgr(certDir, m)
+	tlsConfig.GetClientCertificate = func(*tls.CertificateRequestInfo) (*tls.Certificate, error) { return scm.Current(), nil }
```
Comment (rambohe-ch) [inline pkg/yurthub/certificate/pki.go:107]: maybe `GetClientCertificate` will not used by `yurthub https server`

File: pkg/yurthub/certificate/pki.go [hunk 4]
```diff
@@ -23,14 +23,76 @@ import (
...
+	}
+
+	klog.V(3).Infof("the %s certificate is not valid for server, try to use the old server certificate in %s",
+		projectinfo.GetHubName(), scm.oldCertFile)
+	if scm.oldCert != nil && time.Now().Before(scm.oldCert.Leaf.NotAfter) {
+		return scm.oldCert
+	} else {
+		klog.V(3).Infof("failed to get the old %s server certificate, wait for the new hub certificate", projectinfo.GetHubName())
+	}
+	return nil
```
Comment (rambohe-ch) [inline pkg/yurthub/certificate/pki.go:88]: return &tls.Certificate{Certificate: nil} ?

File: pkg/yurthub/certificate/pki.go (hunk 3 above)
Comment (qclc) [reply pkg/yurthub/certificate/pki.go]: I think it's right, according to the local test result, this function will not be used by yurthub https server, so I delete the tlsConfig.GetClientCertificate, please take review.

File: pkg/yurthub/certificate/pki.go (hunk 4 above)
Comment (qclc) [reply pkg/yurthub/certificate/pki.go]: The return content has changed, please take review.

File: pkg/yurthub/certificate/pki.go [hunk 5]
```diff
@@ -23,14 +23,76 @@ import (
...
+		return scm.oldCert
+	} else {
+		klog.V(3).Infof("failed to get the old %s server certificate, wait for the new hub certificate", projectinfo.GetHubName())
+	}
+	return &tls.Certificate{Certificate: nil}
+}
+
+// GenTLSConfigUseCertMgrAndCertPool generates a TLS configuration
 // using the given certificate manager and x509 CertPool
 func GenTLSConfigUseCertMgrAndCertPool(
```
Comment (rambohe-ch) [inline pkg/yurthub/certificate/pki.go:93]: how about delete `GenUseCertMgrAndTLSConfig` func and remove routine into `GenTLSConfigUseCertMgrAndCertPool `

Comment (qclc) [reply pkg/yurthub/certificate/pki.go]: The code is updated, please take a review.

**Review by rambohe-ch — APPROVED** [review]

**Pull request merged**